test/transaction_tracker.go

```diff
@@ -0,0 +1,193 @@
+	}
+	tr.client = client
+	tr.sub = sub
+	tr.stopCh = make(chan struct{})
+
+	tr.wg.Add(1)
+	go func() {
+		defer tr.wg.Done()
+		err := tr.trackTransactions()
+		if err != nil {
+			fmt.Printf("trackTransactions failed with error: %v\n", err)
+		}
+	}()
+	return nil
+}
+
+// trackTransactions reads new heads from the heads channel and for each head
+// retrieves the block and places the transactions into the processed map. It
+// signals the newTxs condition for every block that contained transactions.
+func (tr *TransactionTracker) trackTransactions() error {
+	for {
+		select {
+		case h := <-tr.heads:
+			b, err := tr.client.BlockByHash(context.Background(), h.Hash())
+			if err != nil {
+				return err
+			}
+			// If we have transactions then process them
+			if len(b.Transactions()) > 0 {
+				tr.newTxs.L.Lock()
+				for _, t := range b.Transactions() {
+					tr.processed[t.Hash()] = b
+				}
+				// signal
+				tr.newTxs.Signal()
+				tr.newTxs.L.Unlock()
+			}
+		case err := <-tr.sub.Err():
+			// Will be nil if closed by calling Unsubscribe()
+			return err
+		case <-tr.stopCh:
+			return nil
+		}
+	}
+}
+
+// AwaitTransactions waits for the transactions listed in hashes to be
+// processed, it will return the ctx.Err() if ctx expires before all the
+// transactions in hashes were processed or ErrStopped if StopTracking is
+// called before all the transactions in hashes were processed.
+func (tr *TransactionTracker) AwaitTransactions(ctx context.Context, hashes []common.Hash) error {
+	hashmap := make(map[common.Hash]struct{}, len(hashes))
+	for i := range hashes {
+		hashmap[hashes[i]] = struct{}{}
+	}
+	tr.newTxs.L.Lock()
+	defer tr.newTxs.L.Unlock()
+	txsFound := make(chan struct{})
+	var exitErr error
+	// This go-routine will either call Signal when the context expires or the
+	// tracker is closed, allowing us to wake from waiting on these events, or
+	// simply exit if the awaited transactions are processed before the context
+	// expires or the tracker is closed.
+	tr.wg.Add(1)
+	go func() {
+		defer tr.wg.Done()
+		// Note that we have already locked tr.newTxs.L in the creating
+		// go-routine and it will only be unlocked when Wait is called from the
+		// creating go-routine, this means Wait must have been called before we
+		// are able to send our signal.
+		select {
+		case <-ctx.Done():
+			tr.newTxs.L.Lock()
+			defer tr.newTxs.L.Unlock()
+			exitErr = ctx.Err()
+			tr.newTxs.Signal()
+		case <-tr.stopCh:
+			tr.newTxs.L.Lock()
+			defer tr.newTxs.L.Unlock()
+			exitErr = errStopped
+			tr.newTxs.Signal()
+		case <-txsFound:
+			return
+		}
+	}()
+
+	for {
+		// Delete processed transactions from hashmap
+		for hash := range hashmap {
+			_, ok := tr.processed[hash]
+			if ok {
+				delete(hashmap, hash)
+			}
+		}
+		// If there are no transactions left then they have all been processed.
+		if len(hashmap) == 0 {
+			close(txsFound) // Signal to the go routine that it can exit
+			return nil
+		}
+		tr.newTxs.Wait()
```

Comment:
If I understand correctly we need the condition vars to deal with the fact that while we can put all the finalized transaction into the shared hashmap `tr.processed` (maybe `tr.seenTransactions`), we only want to scan through it once we get a new block.

What do you think about using a channel to signal a new block event? I think it'd play nicer with ctx.Done() and allow you to get rid of the extra go-routine. The extra channel could be a new sub or a channel  that `trackTransactions` only sends to when `AwaitTransactions` is running (requires a lock, but is doable).

Reply:
I'm going to have a shot at this, getting rid of that go-routine would be nice :)

Reply:
I've made the change, its looking much neater. Dropped a sweet 20 LOC as well :)

With regards to renaming 'processed', I chose that name because I thought it made a clear distinction between transactions that came from finished blocks versus transactions that the tracker had been asked to wait for. With 'seen' I feel like it makes the distinction less clear.